activerecord/lib/active_record/destroy_association_async_job.rb

```diff
@@ -23,9 +23,12 @@ def perform(
         raise DestroyAssociationAsyncError, "owner record not destroyed"
       end
 
-      association_model.where(association_primary_key_column => association_ids).find_each do |r|
-        r.destroy
-      end
+      assoc_pk_cols = Array(association_primary_key_column)
+
+      association_ids
+        .map { |assoc_ids| association_model.where(assoc_pk_cols.zip(Array(assoc_ids)).to_h) }
```

Comment:
If I'm following the code correctly, it may result in an unwanted change of the query being built for a regular use-case like `id: [1,2,3]` passed for example from here - https://github.com/Shopify/rails/blob/b3a54378fa4dba76bf489860c5819d32144120cd/activerecord/lib/active_record/associations/has_many_association.rb#L47

with this change it looks like we will build a query like `WHERE (id=1) OR (id=2) OR (id=3)` which is unnecessary
We should either branch and build the query in two different ways or keep the current code but additionally wrap `association_ids` into an Array for a single-column primary key case. `association_ids` will look like `[[1,2,3]]` and `assoc_pk_cols` will look like `[:id]` 
So `association_ids.map` only iterates once building the same `.where(id: [1,2,3])` query once.

Reply:
Decided to branch them separately in the job